test/base-fork/Morpho.t.sol

```diff
@@ -0,0 +1,272 @@
+
+    address constant MORPHO_VAULT_USDS = 0x0fFDeCe791C5a2cb947F8ddBab489E5C02c6d4F7;
+    address constant MORPHO_VAULT_USDC = 0x305E03Ed9ADaAB22F4A58c24515D79f2B1E2FD5D;
+
+    IERC4626 usdsVault = IERC4626(MORPHO_VAULT_USDS);
+    IERC4626 usdcVault = IERC4626(MORPHO_VAULT_USDC);
+
+    function setUp() public override {
+        super.setUp();
+
+        vm.startPrank(Base.SPARK_EXECUTOR);
+
+        // Add in the idle markets so deposits can be made
+        MarketParams memory usdsParams = MarketParams({
+            loanToken:       Base.USDS,
+            collateralToken: address(0),
+            oracle:          address(0),
+            irm:             address(0),
+            lltv:            0
+        });
+        MarketParams memory usdcParams = MarketParams({
+            loanToken:       Base.USDC,
+            collateralToken: address(0),
+            oracle:          address(0),
+            irm:             address(0),
+            lltv:            0
+        });
+        IMorpho(MORPHO).createMarket(
+            usdsParams
+        );
+        // USDC idle market already exists
+        IMetaMorpho(MORPHO_VAULT_USDS).submitCap(
+            usdsParams,
+            type(uint184).max
+        );
+        IMetaMorpho(MORPHO_VAULT_USDC).submitCap(
+            usdcParams,
+            type(uint184).max
+        );
+
+        skip(1 days);
+
+        IMetaMorpho(MORPHO_VAULT_USDS).acceptCap(usdsParams);
+        IMetaMorpho(MORPHO_VAULT_USDC).acceptCap(usdcParams);
+        
+        Id[] memory supplyQueueUSDS = new Id[](1);
+        supplyQueueUSDS[0] = MarketParamsLib.id(usdsParams);
+        IMetaMorpho(MORPHO_VAULT_USDS).setSupplyQueue(supplyQueueUSDS);
+        Id[] memory supplyQueueUSDC = new Id[](1);
+        supplyQueueUSDC[0] = MarketParamsLib.id(usdcParams);
+        IMetaMorpho(MORPHO_VAULT_USDC).setSupplyQueue(supplyQueueUSDC);
+
+        rateLimits.setRateLimitData(
+            RateLimitHelpers.makeAssetKey(
+                foreignController.LIMIT_VAULT_DEPOSIT(),
+                MORPHO_VAULT_USDS
+            ),
+            25_000_000e18,
+            uint256(5_000_000e18) / 1 days
+        );
+        rateLimits.setRateLimitData(
+            RateLimitHelpers.makeAssetKey(
+                foreignController.LIMIT_VAULT_DEPOSIT(),
+                MORPHO_VAULT_USDC
+            ),
+            25_000_000e6,
+            uint256(5_000_000e6) / 1 days
+        );
+
+        vm.stopPrank();
+    }
+
+    function _getBlock() internal pure override returns (uint256) {
+        return 22841965;  // November 24, 2024
+    }
+
+}
+
+/**********************************************************************************************/
+/*** Only testing USDS failure modes because it is the same for USDC                        ***/
+/**********************************************************************************************/
+
+contract MorphoFailureTests is MorphoBaseTest {
```

Comment:
Break out contracts by function, deposit, withdraw, and redeem failures

Reply:
at a certain point this seems excessive having 1-2 tests per contract